nettoyage.py

```diff
@@ -383,11 +386,49 @@ def data_inputation(df):
 
 
 def replace_char(df):
-    """Fonction pour mettre en qualité le champ objetMarche"""
+    """Fonction pour mettre en qualité le champ objet"""
     # Remplacement brutal du caractère (?) par un espace
     df['objet'] = df['objet'].str.replace('�', 'XXXXX')
     return df
 
+# Pour la PR
+# Cette fonction doit s'insérer après le traitement de la branche Add feature modifications
+
+def regroupement_marche_complet(df):
+    """la colonne id n'est pas unique. Cette fonction permet de la rendre unique en regroupant 
+    les marchés en fonction de leur objets/date de publication des données et montant.
+    Ajoute dans le meme temps la colonne nombreTitulaireSurMarchePresume"""
+    # Creation du sub DF necessaire
+    df_titulaires = pd.DataFrame()
+    df_to_update = pd.DataFrame()
+    df_intermediaire = df[["objet", "datePublicationDonnees", "montant", "id"]] # "datePublicationDonnees",
+    # On regroupe selon l objet du marché. Attention, objet n est pas forcément unique mais idMarche ne l'est pas non plus.
+    df_group = pd.DataFrame(df_intermediaire.groupby(["objet",
+                                                      "datePublicationDonnees", "montant"])["id"])
+    # Initialisation du resultat sous forme de liste
+    # df_group a un multi_index objet-datePublicationDonnees
+    index = df_group.index
+    for i in range(len(df_group)):
+        # dataframe contenant les id d'un meme marche
+        ids_to_modify = df_group[1].iloc[i]
+        # Contient les index des lignes d'un meme marché. Utile pour le update
+        new_index = list(ids_to_modify.index)
+        # Création du dataframe avec id en seule colonne et comme index les index dans le df initial
+        df_avec_bon_id = pd.DataFrame(len(new_index)*[max(ids_to_modify)], index=new_index, columns = ["id"])
+        # Création d'un dataframe intermédiaire avec comme colonne nombreTitulaireSurMarchePresume
+        df_nbtitulaires = pd.DataFrame(len(new_index)*[len(new_index)], index=new_index, columns = ["nombreTitulaireSurMarchePresume"])
+        df_to_update = pd.concat([df_to_update, df_avec_bon_id])
+        # Dataframe permettant de faire la jointure pour ajouter la colonne nombreTitulaireSurMarchePresume dans le df initial
+        df_titulaires = pd.concat([df_titulaires, df_nbtitulaires])
+        # Arbitraire, faire un update a chaque étape rallonge le temps d'execution, un update final idem.
+        if len(df_to_update) > 50000:
+            df.update(df_to_update)
+            df_to_update = pd.DataFrame()
+    # Ajout effectif de nombreTitulaireSurMarchePresume
+    df = df.merge(df_titulaires, how='left', left_index=True, right_index=True)
+    df.update(df_to_update)
```

Comment:
dans le cas d'un `df_to_update > 5000` on fait ligne 425 `df.update(df_to_update)`
et ensuite on remet le dataframe à 0 puis ligne 429 on refait `df.update(df_to_update)` mais avec un df vide, il y à une raison ?

Reply:
ligne 429 on sort de la boucle for 
donc on traite la cas ou df_to_update a une taille inférieur à 5000 en sortie de boucle